cirq-core/cirq/sim/simulator_base.py

```diff
@@ -218,9 +219,10 @@ def _core_iterator(
                 except TypeError:
                     raise TypeError(f"{self.__class__.__name__} doesn't support {op!r}")
 
-            step_state = sim_state.create_merged_state()
-            yield self._create_step_result(step_state, step_state.qubit_map)
-            step_state.log_of_measurement_results.clear()
+            step_result = self._create_step_result(sim_state)
+            yield step_result
+            sim_state = step_result._sim_state
```

Comment:
This is to enable `StepResult.set_state_vector` etc.

The same behavior has existed long before any of these PRs. State is always passed by reference into the step results for obvious performance reasons, so step results change under the hood if you don't explicitly copy them during the simulation.

There are two ways to enable `StepResult.set_state_vector` though.
1. Have `set_state_vector` explicitly create a new sim_state, and then have the simulator explicitly reload it.
2. Have `set_state_vector` update the existing `step_result._sim_state` directly, and then the simulator will implicitly pick that up because they share the reference.

I went with the first option because I think it makes the thing more explicit, and it also allows for a generic `set_sim_state(TOperationTarget)` if we ever want to do that instead of just a `set_state_vector`. The implementation of the second option is a little hairy too, as it either requires a generic `OperationTarget.set_state()`, which I wasn't quite ready to commit to, or some ugly type checking and casting depending on whether sim_state is an ActOnArgs or an ActOnArgsContainer.

Reply:
Given that this is the expected behavior, the current approach seems fine to me.

The fact that this is expected still comes as a bit of a surprise to me, although I can definitely understand the motivation behind it. Do we have tests that explicitly demonstrate this behavior? If not, I think the `simulator_base_test` seems like the appropriate place to put those tests.

Reply:
Added two tests to the bottom of simulator_base_test.

Reply:
test_state_vector_copy is one test that already included checks for this functionality if you wanted to take a look. I also mention that test in the main PR description above.

Reply:
Thanks, these (plus the two you pointed out) provide the coverage I was looking for.